keps/sig-node/2712-pod-priority-based-graceful-node-shutdown/README.md

```diff
@@ -0,0 +1,553 @@
+determine the order in which the pods are stopped.
+
+## Motivation
+
+The [node graceful shutdown KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2000-graceful-node-shutdown) added support to the kubelet to detect
+that a node is shutting down and making sure that the pods are gracefully
+stopped before allowing the shutdown to proceed. 
+
+The feature added flags to specify the total time for shutdown and the
+time to reserve for shutting down critical pods.
+
+However, there is a need to allow more fine grained control over the
+pod shutdown order beyond _critical_ and _regular_ pods.
+
+Also, in general, kubernetes API design discourages hard coding anything by instances
+of name. 
+
+Instead of looking at the pod priority class names, we can instead
+look at the pod priority class values to allow more control over pod
+shutdown order.
+
+### Goals
+
+*   Make the kubelet use shutdown configuration based on pod priority values for
+    graceful shutdown.
+
+### Non-Goals
+
+*   Non-Linux hosts aren't supported
+*   Let users modify or change existing pod lifecycle or introduce new inner
+    pod depencides / shutdown ordering
+*   Provide guarantee to handle all cases of graceful node shutdown, for
+    example abrupt shutdown or sudden power cable pull can’t result in graceful
+    shutdown
+
+## Proposal
+
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+*   As a cluster administrator, I can configure the nodes in my cluster to
+    allocate different graceful shutdown durations for different pod priority
+    value ranges to terminate them gracefully during node shutdown
+
+
+### Implementation
+
+This implementation builds on top of the node graceful shutdown feature
+by introducing additional configuration. A new feature flag called
+`PodPriortityBasedGracefulShutdown` will be added to control the behavior
+of the kubelet.
+
+We will describe the configuration by using an example. Say, the
+following custom pod priority classes are created in a cluster:
+
+|Pod priority class name|Pod priority class value|
+|-----------------------|------------------------|
+|custom-class-a         | 100000                 |
+|custom-class-b         | 10000                  |
+|custom-class-c         | 1000                   |
+|regular/unset          | 0                      |
+
+We could set kubelet configuration to stop the pods as:
+
+|Pod priority class value|Shutdown period|
+|------------------------|---------------|
+| 100000                 |300 seconds    |
+| 10000                  |180 seconds    |
+| 1000                   |120 seconds    |
+| 0                      |60 seconds     |
+
+The above table implies that any pod with priority value >= 100000 will get
+300 seconds to stop, any pod with value >= 10000 and < 100000 will get 180
+seconds to stop, any pod with value >= 1000 and < 10000 will get 120 seconds to stop. 
+Finally, all other pods will get 60 seconds to stop.
```

Comment:
I would illustrate that the shutdown periods don't need to be ordered (assuming that's true)
```suggestion
| 100000                 |10 seconds     |
| 10000                  |180 seconds    |
| 1000                   |120 seconds    |
| 0                      |60 seconds     |

The above table implies that any pod with priority value >= 100000 will get
just 10 seconds to stop, any pod with value >= 10000 and < 100000 will get 180
seconds to stop, any pod with value >= 1000 and < 10000 will get 120 seconds to stop. 
Finally, all other pods will get 60 seconds to stop.
```

Reply:
(this also matches how I'd set up a node - once there are no workload Pods, actually I don't mind if the node shutdown is less than graceful).

Reply:
Yes, they don't have to be ordered.

Reply:
Thanks!